Switch user and drop privileges

joe4dev opened the pull request:
Depends on https://github.com/localstack/lambda-runtime-init/pull/12

Switch to `sbx_user1051` user for runtime parity and drop root privileges.

# Limitations

* `DropPrivileges` breaks debugging after `syscall`. Debugging before the syscall is possible and resuming (at least) a few lines before the syscall works but stepping over the syscall or any breakpoint beyond the syscalls breaks debugging. Hence, we need to disable `DropPrivileges` for debugging:
  * Manually comment `DropPriviledges`
  * Introduce an environment variable to conditionally skip `DropPriviledges`

joe4dev committed:
Switch to sbx user and drop root privileges

joe4dev committed:
Distinguish between empty and unset environment variables

Make `GetenvWithDefault` behave like `os.environ.get` in Python

joe4dev committed:
Make user configurable

joe4dev committed:
Change owner of tmp directory

dominikschubert commented on cmd/localstack/user.go:
```diff
@@ -0,0 +1,111 @@
+	}
+}
+
+// doesFileContainEntry returns true of the entry string is contained in the given file
```
```suggestion
// doesFileContainEntry returns true if the entry string exists in the given file
```

dominikschubert commented on cmd/localstack/user.go:
```diff
@@ -0,0 +1,111 @@
+
+// DropPrivileges switches to another UNIX user by dropping root privileges
+// Initially based on https://stackoverflow.com/a/75545491/6875981
+func DropPrivileges(userToSwitchTo string) {
+	// Lookup user and group IDs for the user we want to switch to.
+	userInfo, err := user.Lookup(userToSwitchTo)
+	if err != nil {
+		log.Errorln("Error looking up user:", userToSwitchTo, err)
+	}
```
```suggestion
func DropPrivileges(userToSwitchTo string) error {
	// Lookup user and group IDs for the user we want to switch to.
	userInfo, err := user.Lookup(userToSwitchTo)
	if err != nil {
		log.Errorln("Error looking up user:", userToSwitchTo, err)
		return err
	}
	...
	return nil
```

This is more of a general feedback for the methods in this file. It would generally be better to "fail" early and bubble up the error when receiving an error we don't plan to recover from.

joe4dev replied:
done for methods without a return value. Should we return a tuple for all methods with return values?

joe4dev replied:
switched to warn logging for recoverable errors for now

dominikschubert commented on cmd/localstack/user.go:
```diff
@@ -0,0 +1,111 @@
+	// Limitation: Debugger gets stuck when stepping over these syscalls!
+	// No breakpoints beyond this point are hit.
+	// Set group ID (real and effective).
+	err = syscall.Setgid(gid)
+	if err != nil {
+		log.Errorln("Failed to set group ID:", err)
+	}
```
```suggestion
	if err = syscall.Setgid(gid); err != nil {
		log.Errorln("Failed to set group ID:", err)
	}
```

dominikschubert replied:
a bit more idiomatic

joe4dev replied:
good to learn. adjusted accordingly
It seems that single returns are typically combined and tuple returns are on a separate line

dominikschubert commented on cmd/localstack/user.go:
```diff
@@ -0,0 +1,111 @@
+	}
+	// Set user ID (real and effective).
+	err = syscall.Setuid(uid)
+	if err != nil {
+		log.Errorln("Failed to set user ID:", err)
+	}
```
same here

dominikschubert commented on cmd/localstack/main.go:
```diff
@@ -60,6 +62,20 @@ func main() {
+		uid := 993
+		gid := 990
+		AddUser(lsOpts.User, uid, gid)
+		err := os.Chown("/tmp", uid, gid)
+		if err != nil {
+			log.Errorln("Error changing owner of /tmp:", err)
+		}
```
```suggestion
		if err := os.Chown("/tmp", uid, gid); err != nil {
			log.Errorln("Error changing owner of /tmp:", err)
		}
```

dominikschubert reviewed (approved):
Minor nits, otherwise LGTM! :rocket:

joe4dev committed:
Apply review feedback

dfangl commented on cmd/localstack/main.go:
```diff
@@ -60,6 +62,19 @@ func main() {
 	log.SetLevel(log.DebugLevel)
 	log.SetReportCaller(true)
 
+	// Switch to non-root user and drop root privileges
+	if IsRootUser() && lsOpts.User != "" {
+		uid := 993
+		gid := 990
+		AddUser(lsOpts.User, uid, gid)
+		if err := os.Chown("/tmp", uid, gid); err != nil {
+			log.Errorln("Error changing owner of /tmp:", err)
+		}
+		UserLogger().Debugln("Process running as root user.")
+		DropPrivileges(lsOpts.User)
+		UserLogger().Debugln("Process running as non-root user.")
+	}
+
```
Not a fan of doing this before we start the DNS. Depending on a docker version (https://github.com/moby/moby/pull/41030), we might not have the permission for binding a port < 1024 later. I would move the privilege drop after the code archive extraction + dns server start.

joe4dev replied:
Makes a lot of sense. Good catch 👍 
Moved it right after `go RunDNSRewriter`

dfangl reviewed (changes requested):
Only problem I have is the location of the privilege dropping, which should, in my opinion, be after binding port 53 and extracting the archives.

joe4dev committed:
Drop privileges after enabling DNS server

As @dfangl pointed out:
Binding port 53 might require root permissions for binding port < 1024 depending on the Docker version
https://github.com/moby/moby/pull/41030

dfangl reviewed (approved):
LGTM now. We should however merge the other PR first, and might want to rebase this one to the `localstack` branch.

joe4dev committed:
Unset LS-internal env variables

joe4dev committed:
Adjust log level

Using warning level for recoverable warnings rather than breaking errors.

dominikschubert reviewed (approved):
LGTM :+1:
